### PR TITLE
Closes #9354 improve resume downloads in slow networks

### PR DESCRIPTION
**Pull request by Amejia481**
In cases where users with slow networks start a download [but quickly press pause and then resume](https://github.com/mozilla-mobile/fenix/issues/9354#issuecomment-730256235), there will be not enough time for bytes to be copied, but the stream in [[download.response] ](https://github.com/mozilla-mobile/android-components/blob/d19d167a48658d10ccd95a53990e368eb16595f1/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L622) will be [closed](https://github.com/mozilla-mobile/fenix/issues/9354#issuecomment-731267368), and when the user will retry we will try to use [[download.response] ](https://github.com/mozilla-mobile/android-components/blob/d19d167a48658d10ccd95a53990e368eb16595f1/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt#L622) over and over with a closed stream as a result the download will failed every time, we have to fallback to httpClient

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt [hunk 1]
```diff
@@ -697,10 +712,20 @@ abstract class AbstractFetchDownloadService : Service() {
 
             outStream.write(data, 0, bytesRead)
         }
+        if (isInStreamClosed) {
+            // In cases where [download.response] is available and users with slow
+            // networks start a download but quickly press pause and then resume
+            // [isResumingDownload] will be false as there will be not enough time
+            // for bytes to be copied, but the stream in [download.response] will be closed,
+            // we have to fallback to [httpClient]
```
Comment (csadilek) [inline components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:720]: Currently this logic would also be applied for downloads already using the HTTP client. I think in this case we don't need to fall back and we don't need to call `performDownload` again as we're already using the same client anyway. I think we should check for that here...only fall back to the fetch client if we're not already using it.

Comment (csadilek) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:720]: I also think it's worth filing this for GV to potentially improve the API.

**Review by csadilek — COMMENTED** [review]
Looks good! I found one thing below I think we should improve before landing.

**Review by csadilek — COMMENTED** [review]
Looks like the commit references the wrong ticket number, should be https://github.com/mozilla-mobile/android-components/issues/9033.

**Comment by Amejia481** [comment]
> Looks like the commit references the wrong ticket number, should be #9033.

Thanks, I will update it with the correct one :)

File: components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt (hunk 1 above)
Comment (Amejia481) [reply components/feature/downloads/src/main/java/mozilla/components/feature/downloads/AbstractFetchDownloadService.kt:720]: > Currently this logic would also be applied for downloads already using the HTTP client. I think in this case we don't need to fall back and we don't need to call performDownload again as we're already using the same client anyway. I think we should check for that here...only fall back to the fetch client if we're not already using it.

👍🏽 

> I also think it's worth filing this for GV to potentially improve the API.

👍🏽

**Review by csadilek — APPROVED** [review]
👍 🚢 

If you have a GV ticket, I would link it in your comment above where you describe why we're doing this.

**Comment by Amejia481** [comment]
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1679852 for it.

**Pull request merged**